File: src/main/scala/net/projecteuler/scala/Problem116.scala
package net.projecteuler.scala

import scala.collection.mutable._

import net.projecteuler.scala.util.Partition

/**
 * Investigating the number of ways of replacing
 * square tiles with one of three colored tiles.
 *
 * @author vuser
 *
 */
object Problem116 {

    /**
     * Solution is based on the partition of a positive integer n.</br>
     * I.e. 5 can be produced as 2 + 1 + 1 + 1. Because of a number can be used
     * more than one in a given partition we have to use permutation with repetition of that number.
     */
    def main(args: Array[String]): Unit = {
        var counter = BigInt(1)
        // Produces all partition of the given number
        val partitions = Partition.partition(50)
        partitions.foreach {
            p => counter = updateCounter(counter, checkPartition(p))
        }
        assert(counter == 20492570929L)
    }

    private def updateCounter(counter: BigInt, map: Map[Int, Int]): BigInt = {
        if (map.isEmpty) {
            return counter
        }
        // Calculates the permutation with repetition of the numbers
        var numerator = BigInt(0)
        var denominator = BigInt(1)
        for ((_, value) <- map) {
            numerator += value
            denominator = denominator * factorial(BigInt(value))
        }
        counter + (factorial(numerator))./(denominator)
    }

    private def checkPartition(partition: Array[Int]): Map[Int, Int] = {
        val map = new HashMap[Int, Int]
        var count = 0
        partition.foreach {
            i => if (updateCaseInRange(i, map)) count += 1
        }

        if (!map.contains(1) || count != partition.length || map.size != 2) {
            return Map.empty[Int, Int]
        }
        map
    }

    /**
     * If the number is in the given range updates the map.
     */
    private def updateCaseInRange(i: Int, map: Map[Int, Int]) = i match {
        case x if (i >= 1 && i < 5) => update(i, map)
        case _ => false
    }

    private def update(i: Int, map: Map[Int, Int]): Boolean = {
        map get i match {
            case None => map += (i -> 1)
            case Some(x) => map(i) = x + 1
        }
        true
    }

    private def factorial(n: BigInt): BigInt = factorial2(n, 1)

    private def factorial2(n: BigInt, result: BigInt): BigInt = if (n == 0) result else factorial2(n - 1, n * result)
}

File: src/main/java/net/projecteuler/java/Problem116.java
package net.projecteuler.java;

import java.math.BigInteger;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import net.projecteuler.java.util.Partition;

/**
 * Investigating the number of ways of replacing
 * square tiles with one of three colored tiles.
 *
 * @author vuser
 *
 */
public class Problem116 {

	public static void main(String ... args) {
		assert solve() == 20492570929L : "failed";
	}

	/**
	 * Solution is based on the partition of a positive integer n.</br>
	 * I.e. 5 can be produced as 2 + 1 + 1 + 1. Because of a number can be used
	 * more than one in a given partition we have to use permutation with repetition of that number.
	 */
	private static long solve() {
		BigInteger counter = BigInteger.ONE;

		// Produces all partition of the given number
		List<List<Integer>> list = Partition.partition(7);
		for (List<Integer> partition : list) {
			Map<Integer, Integer> map = checkPartition(partition);
			if (map.isEmpty()) {
				continue;
			}

			long numerator = 0;
			BigInteger denominator = BigInteger.ONE;

			// Calculates the permutation with repetition of the numbers
			for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
				numerator += entry.getValue();
				denominator = denominator.multiply(factor(entry.getValue()));
			}
			counter = counter.add(factor(numerator).divide(denominator));
		}

		return counter.longValue();
	}

	private static Map<Integer, Integer> checkPartition(List<Integer> partition) {
		int count = 0;
		Map<Integer, Integer> map = new HashMap<Integer, Integer>();
		for (int i : partition) {
			if (i >= 1 && i < 5) {
				Integer value = map.get(i);
				if (value == null) {
					map.put(i, 1);
				} else {
					map.put(i, value + 1);
				}
				++count;
			}
		}
		// List must contain 1 and 2 or 3 or 4
		if (!map.containsKey(1) || count != partition.size() || map.size() != 2) {
			return Collections.emptyMap();
		}

		return map;
	}

	private static BigInteger factor(long n) {
		BigInteger factor = BigInteger.ONE;
		for (long i = 2; i <= n; i++) {
			factor = factor.multiply(BigInteger.valueOf(i));
		}
		return factor;
	}
}